core/src/main/scala/kafka/server/DelayedFetch.scala

```diff
@@ -95,8 +97,12 @@ class DelayedFetch(delayMs: Long,
                 debug("Satisfying fetch %s immediately since it is fetching older segments.".format(fetchMetadata))
                 return forceComplete()
               } else if (fetchOffset.messageOffset < endOffset.messageOffset) {
-                // we need take the partition fetch size as upper bound when accumulating the bytes
-                accumulatedSize += math.min(endOffset.positionDiff(fetchOffset), fetchStatus.fetchInfo.fetchSize)
+                // we take the partition fetch size as upper bound when accumulating the bytes (skip if a throttled partition)
```

Comment:
It seems that we need to handle case C better. If the follower is lagging on old segments and the quota is exceeded, we may return an empty result well before max wait. Returning an empty result early occasionally is fine. However, in this case, it seems that this can happen continuously.